components/web_discovery/browser/content_scraper.cc

```diff
@@ -0,0 +1,379 @@
+#include "base/json/json_writer.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_split.h"
+#include "base/task/thread_pool.h"
+#include "brave/components/web_discovery/browser/patterns.h"
+#include "brave/components/web_discovery/browser/privacy_guard.h"
+#include "brave/components/web_discovery/browser/util.h"
+
+namespace web_discovery {
+
+namespace {
+
+constexpr char kUrlAttrId[] = "url";
+constexpr char kCountryCodeAttrId[] = "ctry";
+constexpr char kFieldsValueKey[] = "fields";
+constexpr char kIdValueKey[] = "id";
+constexpr char kUrlValueKey[] = "url";
+
+constexpr char kRefineSplitFuncId[] = "splitF";
+constexpr char kRefineMaskURLFuncId[] = "maskU";
+constexpr char kRefineParseURLFuncId[] = "parseU";
+constexpr char kRefineJsonExtractFuncId[] = "json";
+
+constexpr char kParseURLQueryExtractType[] = "qs";
+
+std::string RefineSplit(const std::string& value,
+                        const std::string& delimiter,
+                        int index) {
+  auto split = base::SplitStringUsingSubstr(
+      value, delimiter, base::WhitespaceHandling::KEEP_WHITESPACE,
+      base::SPLIT_WANT_ALL);
+  std::string encoded_result;
+  if (index < 0 || static_cast<size_t>(index) >= split.size()) {
+    encoded_result = value;
+  } else {
+    encoded_result = split[index];
+  }
+  return DecodeURLComponent(encoded_result);
+}
+
+std::optional<std::string> RefineParseURL(const std::string& value,
+                                          const std::string& extract_type,
+                                          const std::string& key) {
+  if (extract_type != kParseURLQueryExtractType) {
+    return std::nullopt;
+  }
+  GURL url(value);
+  if (!url.is_valid() || !url.has_query()) {
+    return std::nullopt;
+  }
+  auto query_value = ExtractValueFromQueryString(url.query_piece(), key);
+  return query_value;
+}
+
+std::optional<std::string> RefineJsonExtract(const std::string& value,
+                                             const std::string& path,
+                                             bool extract_objects) {
+  auto parsed = base::JSONReader::Read(value);
+  if (!parsed || !parsed->is_dict()) {
+    return std::nullopt;
+  }
+  const auto* found_value = parsed->GetDict().FindByDottedPath(path);
+  if (!found_value) {
+    return std::nullopt;
+  }
+  if (found_value->is_string()) {
+    return found_value->GetString();
+  }
+  if ((found_value->is_dict() || found_value->is_list()) && !extract_objects) {
+    return std::nullopt;
+  }
+  std::string encoded_value;
+  if (!base::JSONWriter::Write(*found_value, &encoded_value)) {
+    return std::nullopt;
+  }
+  return encoded_value;
+}
+
+}  // namespace
+
+PageScrapeResult::PageScrapeResult(GURL url, std::string id)
+    : url(url), id(id) {}
+PageScrapeResult::~PageScrapeResult() = default;
+
+ContentScraper::ContentScraper(const ServerConfigLoader* server_config_loader)
+    : sequenced_task_runner_(base::ThreadPool::CreateSequencedTaskRunner({})),
```

Comment:
i think we should explicitly state base::TaskPriority::BEST_EFFORT and TaskPriority::SKIP_ON_SHUTDOWN